usr.bin/at/at.1
.\" $NetBSD: at.1,v 1.11 1999/05/22 05:05:11 jwise Exp $
.\" $OpenBSD: at.1,v 1.6 1998/06/05 00:47:46 deraadt Exp $
.\" $FreeBSD: at.man,v 1.6 1997/02/22 19:54:05 peter Exp $
.Dd April 12, 1995
.Dt "AT" 1
.Os
.Sh NAME
.Nm at, batch, atq, atrm
.Nd queue, examine or delete jobs for later execution
.Sh SYNOPSIS
.Nm at
.Op Fl V
.Op Fl q Ar queue
.Op Fl f Ar file
.Op Fl j Ar jobname
.Op Fl mldrbv
.Ar time
.Nm at
.Op Fl V
.Fl c Ar job Op Ar job ...
.Nm atq
.Op Fl V
.Op Fl q Ar queue
.Op Fl v
.Nm atrm
.Op Fl V
.Ar job
.Op Ar job ...
.Nm batch
.Op Fl V
.Op Fl q Ar queue
.Op Fl f Ar file
.Op Fl mv
.Op Ar time
.Sh DESCRIPTION
.Nm At
and
.Nm batch
read commands from standard input or a specified file which
are to be executed at a later time, using
.Xr sh 1 .
.Bl -tag -width indent
.It Nm at
Executes commands at a specified time.
.It Nm atq
Lists the user's pending jobs, unless the user is the superuser.
In that case, everybody's jobs are listed.
.It Nm atrm
Deletes jobs.
.It Nm batch
Executes commands when system load levels permit.  In other words, when
the load average drops below 1.5, or the value specified in the invocation of
.Nm atrun .
.El
.Pp
.Nm At
allows some moderately complex
.Ar time
specifications.  It accepts times of the form
.Ar HHMM
or
.Ar HH:MM
to run a job at a specific time of day.
(If that time is already past, the next day is assumed.)
You may also specify
.Nm midnight ,
.Nm noon ,
or
.Nm teatime
(4pm)
and you can have a time-of-day suffixed with
.Nm AM
or
.Nm PM
for running in the morning or the evening.
You can also say what day the job will be run,
by giving a date in the form
.Ar \%month-name day
with an optional
.Ar year ,
or giving a date of the form
.Ar MMDDYY
or
.Ar MM/DD/YY
or
.Ar DD.MM.YY .
The specification of a date must follow the specification of
the time of day.
You can also give times like
.Op Nm now
.Nm + Ar count \%time-units ,
where the time-units can be
.Nm minutes ,
.Nm hours ,
.Nm days ,
or
.Nm weeks
and you can tell
.Nm at
to run the job today by suffixing the time with
.Nm today
and to run the job tomorrow by suffixing the time with
.Nm tomorrow.
.Pp
For example, to run a job at 4pm three days from now, you would do
.Nm at 4pm + 3 days ,
to run a job at 10:00am on July 31, you would do
.Nm at 10am Jul 31
and to run a job at 1am tomorrow, you would do
.Nm at 1am tomorrow.
.Pp
For both
.Nm at
and
.Nm batch ,
commands are read from standard input or the file specified
with the
.Fl f
option and executed.
The working directory, the environment (except for the variables
.Nm TERM ,
.Nm TERMCAP ,
.Nm DISPLAY
and
.Nm _ )
and the
.Ar umask
are retained from the time of invocation.
An
.Nm at
or
.Nm batch
command invoked from a
.Xr su 1
shell will retain the current userid.
The user will be mailed standard error and standard output from his
commands, if any. Mail will be sent using the command
.Xr sendmail 8 .
If
.Nm at
is executed from a
.Xr su 1
shell, the owner of the login shell will receive the mail.
.Pp
The superuser may use these commands in any case.
For other users, permission to use at is determined by the files
.Pa /var/at/at.allow
and
.Pa /var/at/at.deny .
.Pp
If the file
.Pa /var/at/at.allow
exists, only usernames mentioned in it are allowed to use
.Nm at .
.Pp
If
.Pa /var/at/at.allow
does not exist,
.Pa /var/at/at.deny
is checked, every username not mentioned in it is then allowed
to use
.Nm at .
.Pp
If neither exists, only the superuser is allowed use of
.Nm at .
.Pp
An empty
.Pa /var/at/at.deny
means that every user is allowed use these commands.
This is the default configuration.
.Sh OPTIONS
.Bl -tag -width indent
.It Fl V
Prints the version number to standard error.
.It Fl q Ar queue
Uses the specified queue.
A queue designation consists of a single letter.  Valid queue designations
range from
.Nm a
to
.Nm z
and
.Nm A
to
.Nm Z .
The
.Nm c
queue is the default for
.Nm at
and the
.Nm E
queue for
.Nm batch .
Queues with higher letters run with increased niceness.
If a job is submitted to a queue designated with an uppercase letter, it
is treated as if it had been submitted to batch at that time.
If
.Nm atq
is given a specific queue, it will only show jobs pending in that queue.
.It Fl m
Send mail to the user when the job has completed even if there was no
output.
.It Fl f Ar file
Reads the job from
.Ar file
rather than standard input.
.It Fl j Ar jobname
Sets the job's name.  The job name is saved in the job file and is printed by
.Nm atq .
.It Fl l
Is an alias for
.Nm atq .
.It Fl d
Is an alias for
.Nm atrm .
.It Fl r
Is an alias for
.Nm atrm .
.It Fl b
Is an alias for
.Nm batch .
.It Fl v
For
.Nm atq ,
shows completed but not yet deleted jobs in the queue.  Otherwise
shows the time the job will be executed.
.It Fl c
Cats the jobs listed on the command line to standard output.
.El
.Sh FILES
.Bl -tag -width /var/at/.lockfile -compact
.It Pa /var/at/jobs
Directory containing job files
.It Pa /var/at/spool
Directory containing output spool files
.It Pa /var/run/utmp
Login records
.It Pa /var/at/at.allow
Allow permission control
.It Pa /var/at/at.deny
Deny permission control
.It Pa /var/at/.lockfile
Job-creation lock file.
.El
.Sh SEE ALSO
.Xr cron 8 ,
.Xr nice 1 ,
.Xr umask 2 ,
.Xr sh 1 ,
.Xr sendmail 8 ,
.Xr atrun 8 .
.Sh BUGS
.Pp
If the file
.Pa /var/run/utmp
is not available or corrupted, or if the user is not logged on at the
time
.Nm at
is invoked, the mail is sent to the userid found
in the environment variable
.Nm LOGNAME .
If that is undefined or empty, the current userid is assumed.
.Pp
.Nm At
and
.Nm batch
as presently implemented are not suitable when users are competing for
resources.
If this is the case for your site, you might want to consider another
batch system, such as
.Nm nqs .
.Sh AUTHORS
At was mostly written by Thomas Koenig <ig25@rz.uni-karlsruhe.de>.
The time parsing routines are by David Parsons <orc@pell.chi.il.us>.


usr.bin/at/at.c
/*	$NetBSD: at.c,v 1.13 1999/05/22 05:05:11 jwise Exp $	*/

/*
 *  at.c : Put file into atrun queue
 *  Copyright (C) 1993, 1994  Thomas Koenig
 *
 *  Atrun & Atq modifications
 *  Copyright (C) 1993  David Parsons
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. The name of the author(s) may not be used to endorse or promote
 *    products derived from this software without specific prior written
 *    permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR(S) ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR(S) BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/* System Headers */
#include <sys/types.h>
#include <sys/param.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <ctype.h>
#include <dirent.h>
#include <errno.h>
#include <fcntl.h>
#include <pwd.h>
#include <signal.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <utmp.h>
#include <locale.h>

#if (MAXLOGNAME-1) > UT_NAMESIZE
#define LOGNAMESIZE UT_NAMESIZE
#else
#define LOGNAMESIZE (MAXLOGNAME-1)
#endif

/* Local headers */
#include "at.h"
#include "panic.h"
#include "parsetime.h"
#include "perm.h"
#include "pathnames.h"
#define MAIN
#include "privs.h"

/* Macros */
#define ALARMC 10		/* Number of seconds to wait for timeout */

#define TIMESIZE 50
#define ATJOB_COMMENT "# jobname "

enum { ATQ, ATRM, AT, BATCH, CAT };	/* what program we want to run */

/* File scope variables */
#ifndef lint
#if 0
static char rcsid[] = "$OpenBSD: at.c,v 1.15 1998/06/03 16:20:26 deraadt Exp $";
#else
__RCSID("$NetBSD: at.c,v 1.13 1999/05/22 05:05:11 jwise Exp $");
#endif
#endif

char *no_export[] =
{
	"TERM", "TERMCAP", "DISPLAY", "_"
};
static int send_mail = 0;

/* External variables */

extern char **environ;
int fcreated;
char *namep;
char atfile[FILENAME_MAX];

char *atinput = (char *)0;	/* where to get input from */
char *jobname = 0;		/* name of job */
char atqueue = 0;		/* which queue to examine for jobs (atq) */
char atverify = 0;		/* verify time instead of queuing job */

/* Function declarations */

static void sigc	__P((int));
static void alarmc	__P((int));
static char *cwdname	__P((void));
static int  nextjob	__P((void));
static void writefile	__P((time_t, char));
static void list_jobs	__P((void));
static void process_jobs __P((int, char **, int));

int main __P((int, char **));

/* Signal catching functions */

static void
sigc(signo)
	int signo;
{
	/* If the user presses ^C, remove the spool file and exit. */
	if (fcreated) {
		PRIV_START
		(void)unlink(atfile);
		PRIV_END
	}

	exit(EXIT_FAILURE);
}

static void
alarmc(signo)
	int signo;
{
	/* Time out after some seconds. */
	panic("File locking timed out");
}

/* Local functions */

static char *
cwdname()
{
	/*
	 * Read in the current directory; the name will be overwritten on
	 * subsequent calls.
	 */
	static char path[MAXPATHLEN];

	return (getcwd(path, sizeof(path)));
}

static int
nextjob()
{
	int jobno;
	FILE *fid;

	if ((fid = fopen(_PATH_SEQFILE, "r+")) != NULL) {
		if (fscanf(fid, "%5x", &jobno) == 1) {
			(void)rewind(fid);
			jobno = (1+jobno) % 0xfffff;	/* 2^20 jobs enough? */
			(void)fprintf(fid, "%05x\n", jobno);
		} else
			jobno = EOF;
		(void)fclose(fid);
		return (jobno);
	} else if ((fid = fopen(_PATH_SEQFILE, "w")) != NULL) {
		(void)fprintf(fid, "%05x\n", jobno = 1);
		(void)fclose(fid);
		return (1);
	}
	return (EOF);
}

static void
writefile(runtimer, queue)
	time_t runtimer;
	char queue;
{
	/*
	 * This does most of the work if at or batch are invoked for
	 * writing a job.
	 */
	int jobno;
	char *ap, *ppos;
	const char *mailname;
	struct passwd *pass_entry;
	struct stat statbuf;
	int fdes, lockdes, fd2;
	FILE *fp, *fpin;
	struct sigaction act;
	char **atenv;
	int ch;
	mode_t cmask;
	struct flock lock;

	(void)setlocale(LC_TIME, "");

	/*
	 * Install the signal handler for SIGINT; terminate after removing the
	 * spool file if necessary
	 */
	memset(&act, 0, sizeof act);
	act.sa_handler = sigc;
	sigemptyset(&(act.sa_mask));
	act.sa_flags = 0;

	sigaction(SIGINT, &act, NULL);

	(void)strncpy(atfile, _PATH_ATJOBS, sizeof(atfile) - 1);
	ppos = atfile + strlen(atfile);

	/*
	 * Loop over all possible file names for running something at this
	 * particular time, see if a file is there; the first empty slot at
	 * any particular time is used.  Lock the file _PATH_LOCKFILE first
	 * to make sure we're alone when doing this.
	 */

	PRIV_START

	if ((lockdes = open(_PATH_LOCKFILE, O_WRONLY | O_CREAT, S_IWUSR | S_IRUSR)) < 0)
		perr2("Cannot open lockfile ", _PATH_LOCKFILE);

	lock.l_type = F_WRLCK;
	lock.l_whence = SEEK_SET;
	lock.l_start = 0;
	lock.l_len = 0;

	act.sa_handler = alarmc;
	sigemptyset(&(act.sa_mask));
	act.sa_flags = 0;

	/*
	 * Set an alarm so a timeout occurs after ALARMC seconds, in case
	 * something is seriously broken.
	 */
	sigaction(SIGALRM, &act, NULL);
	alarm(ALARMC);
	fcntl(lockdes, F_SETLKW, &lock);
	alarm(0);

	if ((jobno = nextjob()) == EOF)
	    perr("Cannot generate job number");

	(void)snprintf(ppos, sizeof(atfile) - (ppos - atfile),
	    "%c%5x%8lx", queue, jobno, (unsigned long) (runtimer/60));

	for (ap = ppos; *ap != '\0'; ap++)
		if (*ap == ' ')
			*ap = '0';

	if (stat(atfile, &statbuf) != 0)
		if (errno != ENOENT)
			perr2("Cannot access ", _PATH_ATJOBS);

	/*
	 * Create the file. The x bit is only going to be set after it has
	 * been completely written out, to make sure it is not executed in
	 * the meantime.  To make sure they do not get deleted, turn off
	 * their r bit.  Yes, this is a kluge.
	 */
	cmask = umask(S_IRUSR | S_IWUSR | S_IXUSR);
	if ((fdes = open(atfile, O_CREAT | O_TRUNC | O_WRONLY, S_IRUSR)) == -1)
		perr("Cannot create atjob file");

	if ((fd2 = dup(fdes)) < 0)
		perr("Error in dup() of job file");

	if (fchown(fd2, real_uid, real_gid) != 0)
		perr("Cannot give away file");

	PRIV_END

	/*
	 * We've successfully created the file; let's set the flag so it
	 * gets removed in case of an interrupt or error.
	 */
	fcreated = 1;

	/* Now we can release the lock, so other people can access it */
	lock.l_type = F_UNLCK;
	lock.l_whence = SEEK_SET;
	lock.l_start = 0;
	lock.l_len = 0;
	(void)fcntl(lockdes, F_SETLKW, &lock);
	(void)close(lockdes);

	if ((fp = fdopen(fdes, "w")) == NULL)
		panic("Cannot reopen atjob file");

	/*
	 * Get the userid to mail to, first by trying getlogin(), which reads
	 * /etc/utmp, then from $LOGNAME or $USER, finally from getpwuid().
	 */
	mailname = getlogin();
	if (mailname == NULL && (mailname = getenv("LOGNAME")) == NULL)
		mailname = getenv("USER");

	if ((mailname == NULL) || (mailname[0] == '\0') ||
	    (strlen(mailname) > LOGNAMESIZE) || (getpwnam(mailname) == NULL)) {
		pass_entry = getpwuid(real_uid);
		if (pass_entry != NULL)
			mailname = pass_entry->pw_name;
	}

	if (atinput != NULL) {
		fpin = freopen(atinput, "r", stdin);
		if (fpin == NULL)
			perr("Cannot open input file");
	}
	(void)fprintf(fp,
			"#!/bin/sh\n"
			"# atrun uid=%u gid=%u\n"
			"# mail %*s %d\n",
		real_uid, real_gid,
		LOGNAMESIZE, mailname, send_mail );
	if (jobname) fprintf(fp,"%s%s\n",ATJOB_COMMENT,jobname);

	/* Write out the umask at the time of invocation */
	(void)fprintf(fp, "umask %o\n", cmask);

	/*
	 * Write out the environment. Anything that may look like a special
	 * character to the shell is quoted, except for \n, which is done
	 * with a pair of "'s.  Dont't export the no_export list (such as
	 * TERM or DISPLAY) because we don't want these.
	 */
	for (atenv = environ; *atenv != NULL; atenv++) {
		int export = 1;
		char *eqp;

		eqp = strchr(*atenv, '=');
		if (ap == NULL)
			eqp = *atenv;
		else {
			int i;

			for (i = 0;i < sizeof(no_export) /
			    sizeof(no_export[0]); i++) {
				export = export
				    && (strncmp(*atenv, no_export[i],
					(size_t) (eqp - *atenv)) != 0);
			}
			eqp++;
		}

		if (export) {
			(void)fwrite(*atenv, sizeof(char), eqp - *atenv, fp);
			for (ap = eqp; *ap != '\0'; ap++) {
				if (*ap == '\n')
					(void)fprintf(fp, "\"\n\"");
				else {
					if (!isalnum(*ap)) {
						switch (*ap) {
						case '%': case '/': case '{':
						case '[': case ']': case '=':
						case '}': case '@': case '+':
						case '#': case ',': case '.':
						case ':': case '-': case '_':
							break;
						default:
							(void)fputc('\\', fp);
							break;
						}
					}
					(void)fputc(*ap, fp);
				}
			}
			(void)fputs("; export ", fp);
			(void)fwrite(*atenv, sizeof(char), eqp - *atenv - 1, fp);
			(void)fputc('\n', fp);
		}
	}
	/*
	 * Cd to the directory at the time and write out all the
	 * commands the user supplies from stdin.
	 */
	(void)fputs("cd ", fp);
	for (ap = cwdname(); *ap != '\0'; ap++) {
		if (*ap == '\n')
			fprintf(fp, "\"\n\"");
		else {
			if (*ap != '/' && !isalnum(*ap))
				(void)fputc('\\', fp);

			(void)fputc(*ap, fp);
		}
	}
	/*
	 * Test cd's exit status: die if the original directory has been
	 * removed, become unreadable or whatever.
	 */
	(void)fprintf(fp, " || {\n\t echo 'Execution directory inaccessible' >&2\n\t exit 1\n}\n");

	if ((ch = getchar()) == EOF)
		panic("Input error");

	do {
		(void)fputc(ch, fp);
	} while ((ch = getchar()) != EOF);

	(void)fprintf(fp, "\n");
	if (ferror(fp))
		panic("Output error");

	if (ferror(stdin))
		panic("Input error");

	(void)fclose(fp);


 	PRIV_START

	/*
	 * Set the x bit so that we're ready to start executing
	 */
	if (fchmod(fd2, S_IRUSR | S_IWUSR | S_IXUSR) < 0)
		perr("Cannot give away file");

	PRIV_END

	(void)close(fd2);
	(void)fprintf(stderr, "Job %d will be executed using /bin/sh\n", jobno);
}

static void
list_jobs()
{
	/*
	 * List all a user's jobs in the queue, by looping through
	 * _PATH_ATJOBS, or everybody's if we are root
	 */
	struct passwd *pw;
	DIR *spool;
	struct dirent *dirent;
	struct stat buf;
	struct tm runtime;
	unsigned long ctm;
	char queue;
	int jobno;
	time_t runtimer;
	char timestr[TIMESIZE];
	int first = 1;
	FILE *fp;

	PRIV_START

	if (chdir(_PATH_ATJOBS) != 0)
		perr2("Cannot change to ", _PATH_ATJOBS);

	if ((spool = opendir(".")) == NULL)
		perr2("Cannot open ", _PATH_ATJOBS);

	/* Loop over every file in the directory */
	while ((dirent = readdir(spool)) != NULL) {
		if (stat(dirent->d_name, &buf) != 0)
			perr2("Cannot stat in ", _PATH_ATJOBS);

		/*
		 * See it's a regular file and has its x bit turned on and
		 * is the user's
		 */
		if (!S_ISREG(buf.st_mode)
		    || ((buf.st_uid != real_uid) && !(real_uid == 0))
		    || !(S_IXUSR & buf.st_mode || atverify))
			continue;

		if (sscanf(dirent->d_name, "%c%5x%8lx", &queue, &jobno, &ctm) != 3)
			continue;

		if (atqueue && (queue != atqueue))
			continue;

		runtimer = 60 * (time_t) ctm;
		runtime = *localtime(&runtimer);
		strftime(timestr, TIMESIZE, "%Y-%m-%d %H:%M:%S", &runtime);
		if (first) {
			(void)printf("Date\t\t\tOwner\tQueue\tJob#\tName\n");
			first = 0;
		}
		pw = getpwuid(buf.st_uid);

		(void)printf("%s\t%s\t%c%s\t%d",
		    timestr,
		    pw ? pw->pw_name : "???",
		    queue,
		    (S_IXUSR & buf.st_mode) ? "" : "(done)",
		    jobno);
		fp = fopen(dirent->d_name,"r");
		if (! fp)
		    printf("\t[can't get jobname: %s]\n",strerror(errno));
		else {
		    /* The jobname is the first line beginning with
		       ATJOB_COMMENT.  If a line not beginning with a # is
		       reached, there is no jobname. */
		    int atnl;
		    int c;
		    int x;
		    atnl = 1;
		    while (1) {
			c = getc(fp);
			if (c == EOF) {
			    printf("\n");
			    break;
			}
			if (atnl) {
			    if (c != '#') {
				printf("\n");
				break;
			    } else
				x = 0;
			}
			if (c == '\n')
			    atnl = 1;
			else if (x >= 0) {
			    atnl = 0;
			    if (ATJOB_COMMENT[x]) {
				if (c == ATJOB_COMMENT[x])
				    x ++;
				else
				    x = -1;
			    } else {
				printf("\t");
				while ((c != '\n') && (c != EOF)) {
				    putchar(c);
				    c = getc(fp);
				}
				printf("\n");
				break;
			    }
			}
		    }
		}
	}
	PRIV_END
}

static void
process_jobs(argc, argv, what)
	int argc;
	char **argv;
	int what;
{
	/* Delete every argument (job - ID) given */
	int i;
	struct stat buf;
	DIR *spool;
	struct dirent *dirent;
	unsigned long ctm;
	char queue;
	int jobno;

	PRIV_START

	if (chdir(_PATH_ATJOBS) != 0)
		perr2("Cannot change to ", _PATH_ATJOBS);

	if ((spool = opendir(".")) == NULL)
		perr2("Cannot open ", _PATH_ATJOBS);

	PRIV_END

	/* Loop over every file in the directory */
	while((dirent = readdir(spool)) != NULL) {

		PRIV_START
		if (stat(dirent->d_name, &buf) != 0)
			perr2("Cannot stat in ", _PATH_ATJOBS);
		PRIV_END

		if (sscanf(dirent->d_name, "%c%5x%8lx", &queue, &jobno, &ctm) !=3)
			continue;

		for (i = optind; i < argc; i++) {
			if (atoi(argv[i]) == jobno) {
				if ((buf.st_uid != real_uid) && !(real_uid == 0)) {
					(void)fprintf(stderr,
					    "%s: Not owner\n", argv[i]);
					exit(EXIT_FAILURE);
				}
				switch (what) {
				case ATRM:
					PRIV_START

					if (unlink(dirent->d_name) != 0)
						perr(dirent->d_name);

					PRIV_END

					break;

				case CAT:
					{
						FILE *fp;
						int ch;

						PRIV_START

						fp = fopen(dirent->d_name, "r");

						PRIV_END

						if (!fp)
							perr("Cannot open file");

						while((ch = getc(fp)) != EOF)
							putchar(ch);
					}
					break;

				default:
					(void)fprintf(stderr,
					    "Internal error, process_jobs = %d\n",
					    what);
					exit(EXIT_FAILURE);
					break;
				}
			}
		}
	}
}				/* delete_jobs */

/* Global functions */

int
main(argc, argv)
	int argc;
	char **argv;
{
	int c;
	char queue = DEFAULT_AT_QUEUE;
	char queue_set = 0;
	char *pgm;

	enum {
		ATQ, ATRM, AT, BATCH, CAT
	};				/* what program we want to run */
	int program = AT;		/* our default program */
	char *options = "q:f:mvj:ldbrVc";	/* default options for at */
	int disp_version = 0;
	time_t timer;

	RELINQUISH_PRIVS

	/* Eat any leading paths */
	if ((pgm = strrchr(argv[0], '/')) == NULL)
		pgm = argv[0];
	else
		pgm++;

	namep = pgm;

	/* find out what this program is supposed to do */
	if (strcmp(pgm, "atq") == 0) {
		program = ATQ;
		options = "q:vV";
	} else if (strcmp(pgm, "atrm") == 0) {
		program = ATRM;
		options = "V";
	} else if (strcmp(pgm, "batch") == 0) {
		program = BATCH;
		options = "f:q:mvV";
	}

	/* process whatever options we can process */
	opterr = 1;
	while ((c = getopt(argc, argv, options)) != -1)
		switch (c) {
		case 'v':	/* verify time settings */
			atverify = 1;
			break;

		case 'm':	/* send mail when job is complete */
			send_mail = 1;
			break;

		case 'j':
			jobname = optarg;
			break;

		case 'f':
			atinput = optarg;
			if (! jobname)
			    jobname = optarg;
			break;

		case 'q':	/* specify queue */
			if (strlen(optarg) > 1)
				usage();

			atqueue = queue = *optarg;
			if (!(islower(queue) || isupper(queue)))
				usage();

			queue_set = 1;
			break;

		case 'd':
		case 'r':
			if (program != AT)
				usage();

			program = ATRM;
			options = "V";
			break;

		case 'l':
			if (program != AT)
				usage();

			program = ATQ;
			options = "q:vV";
			break;

		case 'b':
			if (program != AT)
				usage();

			program = BATCH;
			options = "f:q:mvV";
			break;

		case 'V':
			disp_version = 1;
			break;

		case 'c':
			program = CAT;
			options = "";
			break;

		default:
			usage();
			break;
		}
	/* end of options eating */

	if (disp_version)
		(void)fprintf(stderr, "%s version %.1f\n", namep, AT_VERSION);

	if (!check_permission()) {
		(void)fprintf(stderr, "You do not have permission to use %s.\n",
		    namep);
		exit(EXIT_FAILURE);
	}

	/* select our program */
	switch (program) {
	case ATQ:
		if (optind != argc)
			usage();
		list_jobs();
		break;

	case ATRM:
	case CAT:
		if (optind == argc)
			usage();
		process_jobs(argc, argv, program);
		break;

	case AT:
		timer = parsetime(argc, argv);
		if (atverify) {
			struct tm *tm = localtime(&timer);
			(void)fprintf(stderr, "%s\n", asctime(tm));
		}
		writefile(timer, queue);
		break;

	case BATCH:
		if (queue_set)
			queue = toupper(queue);
		else
			queue = DEFAULT_BATCH_QUEUE;

		if (argc > optind)
			timer = parsetime(argc, argv);
		else
			timer = time(NULL);

		if (atverify) {
			struct tm *tm = localtime(&timer);
			(void)fprintf(stderr, "%s\n", asctime(tm));
		}

		writefile(timer, queue);
		break;

	default:
		panic("Internal error");
		break;
	}
	exit(EXIT_SUCCESS);
}
